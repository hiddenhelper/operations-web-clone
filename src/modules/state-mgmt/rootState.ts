import { History } from 'history';

import { IState as IAuthState } from './auth/state';
import { IState as IGeneralState } from './general/state';
import { IState as IClientState } from './client/state';
import { IState as IProjectState } from './project/state';
import { IState as IUserState } from './user/state';
import { IState as IAccessControlSystemState } from './access-control-system/state';
import { IState as IBadgePrintingSystemState } from './badge-printer-system/state';
import { IState as IWorkerState } from './worker/state';
import { IState as ICertificationState } from './certification/state';
import { IState as IFileState } from './file/state';
import { IState as IBadgeState } from './badge/state';
import { IState as IStatisticsState } from './statistics/state';
import { IState as ITrainingState } from './training/state';
import { IState as IPaymentState } from './payment/state';
import { IState as IInvoiceState } from './invoice/state';
import { ApiService } from '../services/ApiService';
import { AuthService } from '../services/AuthService';
import { PrinterService } from '../services/PrinterService';
import { FileService } from '../services/FileService';
import { Logger } from '../services/Logger';
import { IProcoreState } from './procore/state';
import { IState as IProjectNewState } from './project-new/state';

export interface IAction {
  type: string;
  payload: any;
}

export interface IRootState {
  auth: IAuthState;
  user: IUserState;
  client: IClientState;
  project: IProjectState;
  accessControlSystem: IAccessControlSystemState;
  badgePrinterSystem: IBadgePrintingSystemState;
  worker: IWorkerState;
  certification: ICertificationState;
  training: ITrainingState;
  invoice: IInvoiceState;
  file: IFileState;
  badge: IBadgeState;
  statistics: IStatisticsState;
  payment: IPaymentState;
  general: IGeneralState;
  router?: { location: { pathname: string; search: string; hard: string; key: string }; action: string };
  procore: IProcoreState;
  projectNew: IProjectNewState;
}

export interface IEpicDependencies {
  history: History;
  apiService: ApiService;
  authService: AuthService;
  printerService: PrinterService;
  fileService: FileService;
  logger: typeof Logger;
}
