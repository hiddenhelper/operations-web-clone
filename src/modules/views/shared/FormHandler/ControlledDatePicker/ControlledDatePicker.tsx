import React, { memo, useEffect, useCallback, useState, useRef } from 'react';
import { MuiPickersUtilsProvider } from '@material-ui/pickers';
import MomentUtils from '@date-io/moment';

import { CalendarIcon } from '../../../../../constants/';
// import { useHideScroll } from '../../../../../utils/useHideScroll';
import { noop } from '../../../../../utils/generalUtils';
import DatePicker from 'react-datepicker';
import 'react-datepicker/dist/react-datepicker.css';
import { TextField, InputAdornment } from '@material-ui/core';

export interface IControlledDatePickerProps {
  name: string;
  value: string;
  variant: 'standard' | 'outlined' | 'filled';
  icon?: React.ReactNode | null;
  iconPosition?: 'start' | 'end' | null;
  startDate?: string;
  endDate?: string;
  autoAccept?: boolean;
  placeholder?: string;
  styleClass?: string;
  inputProps?: any;
  fullWidth?: boolean;
  style?: any;
  error: boolean;
  required?: boolean;
  onChange: (data: any) => void;
  onOpen?: () => void;
  onAccept?: () => void;
  onClose?: () => void;
  onFocus?: () => void;
  onBlur?: () => void;
  invalidDateMessage?: string;
}
export interface ICustomInputProps {
  onChange?: () => void;
  placeholder?: string;
  value?: string;
  onClick?: () => void;
}

const ControlledDatePicker = ({
  name,
  value,
  variant,
  icon,
  iconPosition,
  startDate,
  endDate,
  required,
  autoAccept = true,
  placeholder = '',
  styleClass = '',
  error,
  fullWidth,
  onChange,
  ...rest
}: IControlledDatePickerProps) => {
  const [state, setState] = useState({ startDate, endDate });
  const ref = useRef(null);
  // const { setHideScroll } = useHideScroll();
  const onDateChange = useCallback(
    date => {
      onChange({
        persist: noop,
        target: {
          name: name,
          value: date?.toUTCString(),
        },
      });
    },
    [onChange, name]
  );
  const extraProps = {};
  if (state.startDate) (extraProps as any).minDate = new Date(state.startDate);
  if (state.endDate) (extraProps as any).maxDate = new Date(state.endDate);

  // const onOpen = useCallback(() => {
  //   setHideScroll(true);
  // }, [setHideScroll]);
  // const onClose = useCallback(() => {
  //   setHideScroll(false);
  // }, [setHideScroll]);

  useEffect(() => {
    setState({ startDate, endDate });
  }, [startDate, endDate]);

  return (
    <>
      <MuiPickersUtilsProvider utils={MomentUtils}>
        {/* <KeyboardDatePicker
          data-testid="keyboard-date-picker"
          format="MM/DD/YYYY"
          mask="___, ___ __, ____"
          refuse={/[^a-z\d]+/gi}
          value={value}
          onChange={onDateChange}
          animateYearScrolling={true}
          inputVariant={variant}
          variant="inline"
          className={styleClass}
          keyboardIcon={icon ? icon : <CalendarIcon />}
          InputAdornmentProps={{ position: iconPosition ? iconPosition : 'start' }}
          autoOk={autoAccept}
          placeholder={placeholder}
          error={error}
          required={required}
          fullWidth={fullWidth}
          onOpen={onOpen}
          onClose={onClose}
          {...extraProps}
          {...rest}
        /> */}
        <DatePicker
          selected={value ? new Date(value) : null}
          onChange={onDateChange}
          placeholderText={placeholder}
          required={required}
          {...extraProps}
          {...rest}
          showMonthDropdown={true}
          showYearDropdown={true}
          dropdownMode="select"
          onCalendarClose={() => ref.current.blur()}
          customInput={
            <TextField
              inputRef={ref}
              data-testid="keyboard-date-picker"
              value={value}
              onChange={onDateChange}
              className={styleClass}
              variant={variant}
              InputProps={{
                startAdornment: (
                  <InputAdornment position="start">
                    <CalendarIcon />
                  </InputAdornment>
                ),
              }}
            />
          }
        />
      </MuiPickersUtilsProvider>
    </>
  );
};

export default memo(ControlledDatePicker);
