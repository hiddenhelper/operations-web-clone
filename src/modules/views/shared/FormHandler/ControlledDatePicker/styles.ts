import { createStyles, makeStyles } from '@material-ui/core/styles';
import { STYLE } from '../../../../../constants';
import { toREM } from '../../../../../utils/generalUtils';

export const useStyles = makeStyles(theme =>
  createStyles({
    datePicker: {
      backgroundColor: STYLE.COLOR.INPUT_BACKGROUND,
      '& .MuiButtonBase-root': {
        outline: 'none',
        '& svg': {
          '& path': {
            fill: STYLE.COLOR.INACTIVE,
          },
        },
      },
      '& p': {
        color: `${STYLE.COLOR.ERROR} !important`,
        fontWeight: 600,
        fontSize: `${toREM(13)}`,
        lineHeight: '21px',
        position: 'absolute',
        top: '49px',
        width: '100%',
        marginLeft: 0,
        marginRight: 0,
      },
    },
    blackIcon: {
      '& .MuiButtonBase-root': {
        outline: 'none',
        '& svg': {
          '& path': {
            fill: `${STYLE.COLOR.BLACK} !important`,
          },
        },
      },
      '& .MuiIconButton-root': {
        padding: '8px',
      },
      '& .MuiOutlinedInput-input': {
        marginLeft: '0px',
        padding: 0,
      },
    },
    fullWidth: { width: '100%' },
    adornmentStart: {
      '& .MuiInputAdornment-positionStart': {
        marginRight: 0,
        marginLeft: '2px',
      },
      '& .MuiOutlinedInput-adornedStart': {
        paddingLeft: 0,
      },
    },
    adornmentEnd: {
      '& .MuiOutlinedInput-adornedEnd': {
        paddingRight: 0,
      },
    },
    leftInput: {
      '& .MuiInputBase-input': {
        paddingLeft: 0,
      },
    },
    baseInput: {
      '& .MuiInputBase-input': {
        height: '40px',
        paddingTop: 0,
        paddingBottom: 0,
        paddingLeft: 0,
      },
    },
    timePicker: {
      backgroundColor: STYLE.COLOR.INPUT_BACKGROUND,
      '& .MuiButtonBase-root': {
        outline: 'none',
        '& svg': {
          '& path': {
            fill: STYLE.COLOR.INACTIVE,
          },
        },
      },
      '& p': {
        color: `${STYLE.COLOR.ERROR} !important`,
        fontWeight: 600,
        fontSize: `${toREM(13)}`,
        lineHeight: '21px',
        position: 'absolute',
        top: '49px',
        width: '100%',
        marginLeft: 0,
        marginRight: 0,
      },
    },
  })
);
