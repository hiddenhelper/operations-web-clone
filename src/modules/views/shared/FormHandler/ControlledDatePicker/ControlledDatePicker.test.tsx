import React from 'react';
import { render, fireEvent, act } from '@testing-library/react';

import ControlledDatePicker, { IControlledDatePickerProps } from './ControlledDatePicker';
import { createMockStore } from 'redux-test-utils';
import { getInitialState } from '../../../../../test/rootState';
import { Provider } from 'react-redux';

jest.mock('@material-ui/pickers', () => ({
  MuiPickersUtilsProvider: ({ children }) => children,
  KeyboardDatePicker: ({ onChange, value }) => {
    return (
      <div>
        <input data-testid="datepicker-item" value={value} onChange={onChange} />
      </div>
    );
  },
}));

describe('ControlledDatePicker', () => {
  let props: IControlledDatePickerProps;

  beforeEach(() => {
    props = {
      name: 'date-picker',
      value: '2020/10/10',
      variant: 'standard',
      onChange: jest.fn(),
    };
  });

  it('should render', () => {
    const wrapper = render(
      <Provider store={createMockStore(getInitialState()) as any}>
        <ControlledDatePicker {...props} />
      </Provider>
    );
    expect(wrapper.container).toMatchSnapshot();
  });

  it('should render past dates of startDate disabled', () => {
    props.startDate = '2020/09/10';
    const { container } = render(
      <Provider store={createMockStore(getInitialState()) as any}>
        <ControlledDatePicker {...props} />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  // it('should onChange', () => {
  //   const wrapper = render(
  //     <Provider store={createMockStore(getInitialState()) as any}>
  //       <ControlledDatePicker {...props} />
  //     </Provider>
  //   );

  //   act(() => {
  //     fireEvent.change(wrapper.getByTestId('datepicker-item'), { target: { value: '2020/10/12' } });
  //   });

  //   expect(props.onChange).toHaveBeenCalled();
  // });
});
