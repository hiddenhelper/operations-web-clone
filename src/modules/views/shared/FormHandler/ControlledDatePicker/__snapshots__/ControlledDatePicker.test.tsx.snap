// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlledDatePicker should render 1`] = `
<div>
  <div
    class="react-datepicker-wrapper"
  >
    <div
      class="react-datepicker__input-container"
    >
      <div
        class="MuiFormControl-root MuiTextField-root"
        data-testid="keyboard-date-picker"
      >
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedStart"
        >
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionStart"
          >
            <svg>
              calendar.svg
            </svg>
          </div>
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedStart"
            placeholder=""
            type="text"
            value="10/10/2020"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ControlledDatePicker should render past dates of startDate disabled 1`] = `
<div>
  <div
    class="react-datepicker-wrapper"
  >
    <div
      class="react-datepicker__input-container"
    >
      <div
        class="MuiFormControl-root MuiTextField-root"
        data-testid="keyboard-date-picker"
      >
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedStart"
        >
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionStart"
          >
            <svg>
              calendar.svg
            </svg>
          </div>
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedStart"
            placeholder=""
            type="text"
            value="10/10/2020"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
